### ISSUE TITLE
Spurious "index signature incompatible" error for function returning an object literal expression

### ISSUE THREAD

**Issue by NoelAbrahams** [microsoft/TypeScript#862]
Hi,

**TypeScript Version**: 1.1.0.1

Here is the repro:

``` TypeScript

interface Foo {
    [index: string]: any;
    name: string;       
}

// Case 1: no error
var f1: Foo = {
        name: 'FooOne',
};

// Case 2: no error 
var f2 = function(): Foo {
    return {
        name: 'FooTwo',
    };
}

// Case 3:
// Compiler Error: error TS2322: Build: Type '{ name: string; }' is not 
   // assignable to type 'Foo'
// IDE Error: Cannot convert '{ name: string; }' to 'Foo':  Index signatures 
   // of types '{ name: string; }' and 'Foo' are incompatible.
var f3 = (): Foo => ({
        name: 'FooThree',
});
```

The expected result is there to be no error, since the object literal in the last case is indexable by `string` to return `any`.
